homeassistant/components/bluetooth_tracker/device_tracker.py

```diff
@@ -65,99 +72,107 @@ def discover_devices(device_id: int) -> List[Tuple[str, str]]:
+    track_new: bool = config_entry.options[CONF_TRACK_NEW]
+    _LOGGER.debug("Tracking new devices is set to %s", track_new)
+
+    update_bluetooth_lock = asyncio.Lock()
+    entities = load_initial_entities(hass, config_entry, async_add_entities)
 
     if request_rssi:
         _LOGGER.debug("Detecting RSSI for devices")
 
+    def see_device(mac: str, device_name: str, rssi=None) -> None:
+        """Mark a device as seen."""
+        attributes = {}
+        if rssi is not None:
+            attributes["rssi"] = rssi
+
+        device = BluetoothEntity(config_entry, mac, device_name, attributes)
+        if device in entities:
+            async_dispatcher_send(hass, TRACKER_UPDATE, mac, device_name, attributes)
+            return
+
+        entities.add(device)
+        async_add_entities([device])
+
     async def perform_bluetooth_update():
         """Discover Bluetooth devices and update status."""
-
         _LOGGER.debug("Performing Bluetooth devices discovery and update")
-        tasks = []
+
+        tracked_devices = set()
 
         try:
             if track_new:
                 devices = await hass.async_add_executor_job(discover_devices, device_id)
                 for mac, device_name in devices:
-                    if mac not in devices_to_track and mac not in devices_to_not_track:
-                        devices_to_track.add(mac)
-
-            for mac in devices_to_track:
-                device_name = await hass.async_add_executor_job(lookup_name, mac)
+                    device = next((d for d in entities if d.unique_id == mac), None)
+                    if not device:
+                        device = BluetoothEntity(config_entry, mac, device_name, {})
+
+                    rssi = None
+                    if request_rssi:
+                        client = BluetoothRSSI(device.mac)
+                        rssi = await hass.async_add_executor_job(client.request_rssi)
+                        client.close()
+
+                    see_device(device.mac, device.device_name, rssi)
+                    tracked_devices.add(device)
+
+            # Only lookup names for un-checked devices
+            entities_to_update = entities - tracked_devices
+            for device in entities_to_update:
+                device_name = await hass.async_add_executor_job(lookup_name, device.mac)
                 if device_name is None:
                     # Could not lookup device name
                     continue
 
                 rssi = None
                 if request_rssi:
-                    client = BluetoothRSSI(mac)
+                    client = BluetoothRSSI(device.mac)
                     rssi = await hass.async_add_executor_job(client.request_rssi)
                     client.close()
 
-                tasks.append(see_device(hass, async_see, mac, device_name, rssi))
-
-            if tasks:
-                await asyncio.wait(tasks)
+                see_device(device.mac, device_name, rssi)
```

Comment:
We should see all current device just once. It's enough to do it above. Looking up the name for new devices should be done before that.